wsw/fingerprint.py

```diff
@@ -4,120 +4,114 @@
-        return spec
-
     def show(self):
         plt.spy(self.fingerprint, origin='lower', aspect='auto')
         plt.show()
 
+    @staticmethod
+    def get_prints(signal, sr, n_fft):
+        signal = Fingerprint._lpfilter(signal, sr)
+        signal = Fingerprint._downsample(signal, sr, new_sr=11025)
+        spec = Fingerprint.stft(signal, n_fft)
+        spec = Fingerprint.denoise(spec)
+        spec = Fingerprint.spec_filter(spec, 6)
+        return spec
+
+    @staticmethod
+    def stft(signal, N):
+        """compute the STFT/spectrogram of given signal with Hamming Window"""
+        spec = librosa.stft(signal, n_fft=N, window=sig.windows.hamming)
+        return np.abs(spec)
+
+    @staticmethod
+    def ft_filter(ft, n_bins, alpha=1):
+        """
+        Filters out weaker frequencies from a fourier transform. Could be thought of as analogous to
+        using Scipy.signal.find_peaks, then converting all non peak data to zeros.
+
+        Peak data is determined by first dividing the frequency bins into logarithmic bands. Then, the
+        strongest frequency from each band is identified. Finally, these strongest frequencies are
+        averaged. All frequencies that fall below this average times a constant, alpha, are converted
+        to zeros.
+
+        :param ft: The fourier transform to filter.
+        :param n_bins: The number of logarithmic bands to divide the frequency domain into
+        :param alpha: Coefficient for determining the threshold at which data is dropped.
+        """
+        binned_ft = Fingerprint._log_bin(ft, n_bins)
+        for i, band in enumerate(binned_ft):
+            band_max = np.max(band)
+            filtered = np.where(band < band_max, 0, band)
+            binned_ft[i] = filtered
+        joined_filtered = np.concatenate(binned_ft)
+        thresh = np.mean(joined_filtered)
+
+        return np.where(joined_filtered < thresh * alpha, 0, joined_filtered)
+
+    @staticmethod
+    def spec_filter(spec, n_bins):
+        """
+        Iteratively filter the DFTs that make up a spectrogram, leaving a sparse matrix
+        representing the strongest frequencies over the time domain. See `Fingerprint.ft_filter`
+        for how this is achieved.
+
+        :param spec: The spectrogram to filter.
+        :param n_bins: The number logarithmic bands to divide the frequency domain into.
+        """
+        filtered = np.zeros_like(spec.T)
+        for i, dft in enumerate(spec.T):
+            filtered[i] = Fingerprint.ft_filter(dft, n_bins)
+        return filtered.T
+
+    @staticmethod
+    def _lpfilter(signal, sr):
+        """Helper method. Applies Low-pass filter that attenuates at 10kHz"""
+        cutoff = 10e3
+        sos = sig.butter(10, cutoff, fs=sr, btype='lowpass', analog=False, output='sos')
+        return sig.sosfilt(sos, signal)
+
+    @staticmethod
+    def _log_bin(arr, n_bins):
+        """Helper method. Divide spectrogram frequency bins logarithmically"""
```

Comment:
Proper docstring - however if these are helper methods and are only being used internally, consider placing these methods within the scope of the function that is using them.  You don't need to expose these to the user.  For example, `_log_bin` is only used once within a function.

As a thought experiment, go through your class and see how often you foresee using a particular method in your class.  If you see it only being used once, create a private function inside the scope of where you're using it and just use it there.  You also do not need to test private functions as they are no longer accessible by the user.